pkg/kernel/Dockerfile

```diff
@@ -1,33 +1,79 @@
 # This file must be kept as much in sync with pkg/new-kernel/Dockerfile as posisble
-FROM lfedge/eve-alpine:145f062a40639b6c65efa36bed1c5614b873be52 AS kernel-build-base
-
-ENV BUILD_PKGS \
-    argp-standalone automake bash bc binutils-dev bison build-base \
-    diffutils flex git gmp-dev gnupg installkernel kmod elfutils-dev    \
-    linux-headers libunwind-dev mpc1-dev mpfr-dev ncurses-dev findutils \
-    openssl-dev patch rsync sed squashfs-tools tar xz xz-dev zlib-dev openssl \
-    attr-dev autoconf file coreutils libtirpc-dev libtool util-linux-dev
+# use the same set of packages for simplicity
+ARG BUILD_PKGS_BASE="argp-standalone automake bash bc binutils-dev bison build-base \
+                     diffutils flex git gmp-dev gnupg installkernel kmod elfutils-dev    \
+                     linux-headers libunwind-dev mpc1-dev mpfr-dev ncurses-dev findutils \
+                     openssl-dev patch rsync sed squashfs-tools tar xz xz-dev zlib-dev openssl \
+                     attr-dev autoconf file coreutils libtirpc-dev libtool util-linux-dev"
+
+FROM lfedge/eve-alpine:1fc941603ddc153f26104865db242e644c16aa99 as kernel-build-base-native
+ARG BUILD_PKGS_BASE
+RUN BUILD_PKGS="${BUILD_PKGS_BASE}" eve-alpine-deploy.sh
+
+# hadolint ignore=DL3029
+FROM --platform=${BUILDPLATFORM} lfedge/eve-alpine:1fc941603ddc153f26104865db242e644c16aa99 as kernel-build-base-cross
+ARG BUILD_PKGS_BASE
+RUN BUILD_PKGS="${BUILD_PKGS_BASE}" eve-alpine-deploy.sh
+
+# hadolint ignore=DL3029
+FROM --platform=${BUILDPLATFORM} lfedge/eve-cross-compilers:706fa9f570a366dc8bfc354504ed9547188101a2 AS cross-compilers
+
+# will use several packages from target arch and copy them to sysroot
+FROM lfedge/eve-alpine:1fc941603ddc153f26104865db242e644c16aa99 AS cross-compile-libs
+ENV PKGS musl-dev libgcc
 RUN eve-alpine-deploy.sh
 
+FROM kernel-build-base-cross AS kernel-cross-target-arm64
+ENV TARGET_ARCH=aarch64
+
+# install cross-compile packages and libs for target
+# hadolint ignore=DL3006
+FROM kernel-cross-target-${TARGETARCH} AS kernel-cross-build-amd64
+ENV BUILD_ARCH=x86_64
+ENV CROSS_COMPILE_ENV="${TARGET_ARCH}"-alpine-linux-musl-
+COPY --from=cross-compilers /packages /packages
+# hadolint ignore=DL3018
+RUN apk add --no-cache --allow-untrusted -X /packages build-base-"${TARGET_ARCH}"
+COPY --from=cross-compile-libs /out/ /usr/"${TARGET_ARCH}"-alpine-linux-musl/
+
+# support cross compile from amd64 for now
+# hadolint ignore=DL3006
+FROM kernel-cross-build-${BUILDARCH} AS kernel-target-arm64-build-amd64
+
+# set host and target to target for the reset of stages (native om emulated)
+FROM kernel-build-base-native AS kernel-target-amd64-build-amd64
```

Comment:
You have all these variables already available from docker:

```
BUILDPLATFORM — matches the current machine. (e.g. linux/amd64)

BUILDOS — os component of BUILDPLATFORM, e.g. linux

BUILDARCH — e.g. amd64, arm64, riscv64

BUILDVARIANT — used to set ARM variant, e.g. v7

TARGETPLATFORM — The value set with --platform flag on build

TARGETOS - OS component from --platform, e.g. linux

TARGETARCH - Architecture from --platform, e.g. arm64
```
It isn't possible to simplify these stuff here?

Reply:
The problem is about arm64->aarch64 conversion. But I tried to simplify the logic by moving `BUILD_ARCH` and `TARGET_ARCH` into eve-alpine (thanks @Insei for the idea) inside https://github.com/lf-edge/eve/pull/2978/commits/0c0743aad70354db8d15a5ed2a74b503df70a6bc. So now we set only few ENVs and most of the logic is about chaining from cross or native base image.

Reply:
> The problem is about arm64->aarch64 conversion. But I tried to simplify the logic by moving BUILD_ARCH and TARGET_ARCH into eve-alpine (thanks @Insei for the idea) inside https://github.com/lf-edge/eve/commit/0c0743aad70354db8d15a5ed2a74b503df70a6bc.

That is a _really_ nice idea. Simple and elegant, and everything gets it.

I will ping the buildkit folks about this too.